### PR TITLE
PlatformDefs.h remove _MEMORYSTATUSEX struct

### PR DESCRIPTION
**Pull request by lrusak**
More PlatformDefs.h cleanup.

1) I started by namespacing the three methods in `XMemUtils.h` and then created a wrapper for these functions for windows.
2) I removed the windows specific `_MEMORYSTATUSEX` struct from PlatformDefs.h

I also cleaned up the linux memory method as we don't need to parse `/proc/meminfo`

I'm not sure if this is the best way (by just creating static methods in a namespace that share a common header) but I didn't see another way.

EDIT: I should also note that each commit builds individual (in sequence) but I only tested this on Linux and I don't expect them to build on windows in the same way. So if it's important to keep the build without breaking for each commit then I can squash them at the end,

### CODE REVIEW

File: xbmc/platform/win10/CMakeLists.txt [hunk 1]
```diff
@@ -6,7 +6,8 @@ set(SOURCES input/RemoteControlXbox.cpp
             ../win32/dxerr.cpp
             ../win32/Filesystem.cpp
             ../win32/pch.cpp
-            ../win32/WIN32Util.cpp)
+            ../win32/WIN32Util.cpp
+            ../win32/MemUtils.cpp)
```
Comment (fuzzard) [inline xbmc/platform/win10/CMakeLists.txt]: Alphabetical order?

File: xbmc/utils/MemUtils.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+#include <cstdint>
+#include <memory>
+
+namespace KODI
+{
+namespace MEMORY
+{
+  struct MemoryStatus
+  {
+    unsigned int length;
```
Comment (pkerling) [inline xbmc/utils/MemUtils.h]: Uh what is this used for? Why do we need it?

**Review by yol — COMMENTED** [review]
Looking good, but I totally don't get why you kept the `length` field

File: xbmc/utils/MemUtils.h (hunk 2 above)
Comment (lrusak) [reply xbmc/utils/MemUtils.h]: I just kept the same windows idioms. I can remove it though now that you pointed it out :)

https://docs.microsoft.com/en-gb/windows/desktop/api/sysinfoapi/nf-sysinfoapi-globalmemorystatusex

Comment (pkerling) [reply xbmc/utils/MemUtils.h]: Yes, please remove :3

**Comment by notspiff** [comment]
rebase and squash

File: xbmc/platform/posix/MemUtils.cpp [hunk 3]
```diff
@@ -69,10 +69,10 @@ void GetMemoryStatus(MemoryStatus* buffer)
   uint64_t physmem;
   size_t len = sizeof physmem;
   std::array<int, 2> mib =
-  {
+  {{
     CTL_HW,
     HW_MEMSIZE,
-  };
+  }};
```
Comment (Rechi) [inline xbmc/platform/posix/MemUtils.cpp]: There should at least added a doxygen comment about the compiler bug, so the workaround can be found and removed later.
You could also disable the warning for this lines via a pragma.

```
#if defined(__apple_build_version__) && __apple_build_version__ < 10000000
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wmissing-braces"
#endif
  std::array<int, 2> abc =
  {
    CTL_HW,
    HW_MEMSIZE,
  };
#if defined(__apple_build_version__) && __apple_build_version__ < 10000000
#pragma clang diagnostic pop
#endif
```

File: xbmc/guilib/Texture.cpp [hunk 4]
```diff
@@ -94,12 +92,12 @@ void CBaseTexture::Allocate(unsigned int width, unsigned int height, unsigned in
...
   CLAMP(m_imageHeight, m_textureHeight);
 
-  _aligned_free(m_pixels);
+  KODI::MEMORY::AlignedFree(m_pixels);
   m_pixels = NULL;
   if (GetPitch() * GetRows() > 0)
   {
     size_t size = GetPitch() * GetRows();
-    m_pixels = (unsigned char*) _aligned_malloc(size, 32);
+    m_pixels = (unsigned char*) KODI::MEMORY::AlignedMalloc(size, 32);
```
Comment (Rechi) [inline xbmc/guilib/Texture.cpp]: use C++ style cast

File: xbmc/interfaces/legacy/ModuleXbmc.cpp [hunk 5]
```diff
@@ -251,10 +248,9 @@ namespace XBMCAddon
     long getFreeMem()
     {
       XBMC_TRACE;
-      MEMORYSTATUSEX stat;
-      stat.dwLength = sizeof(MEMORYSTATUSEX);
-      GlobalMemoryStatusEx(&stat);
-      return (long)(stat.ullAvailPhys  / ( 1024 * 1024 ));
+      KODI::MEMORY::MemoryStatus stat;
+      KODI::MEMORY::GetMemoryStatus(&stat);
+      return (long)(stat.availPhys  / ( 1024 * 1024 ));
```
Comment (Rechi) [inline xbmc/interfaces/legacy/ModuleXbmc.cpp]: use C++ style cast

File: xbmc/platform/posix/MemUtils.cpp [hunk 6]
```diff
@@ -33,44 +49,44 @@ void *_aligned_malloc(size_t s, size_t alignTo) {
...
+  {
+    CTL_HW,
+    HW_MEMSIZE,
+  };
 
   // Total physical memory.
-  if (sysctl(mib, miblen, &physmem, &len, NULL, 0) == 0 && len == sizeof (physmem))
-      lpBuffer->ullTotalPhys = physmem;
+  if (sysctl(mib.data(), mib.size(), &physmem, &len, NULL, 0) == 0 && len == sizeof (physmem))
+      buffer->totalPhys = physmem;
```
Comment (Rechi) [inline xbmc/platform/posix/MemUtils.cpp]: wrong indention

File: xbmc/utils/MemUtils.h [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+#pragma once
+
+#include <cstdint>
+#include <memory>
+
+namespace KODI
+{
+namespace MEMORY
+{
+  struct MemoryStatus
```
Comment (Rechi) [inline xbmc/utils/MemUtils.h]: do not indent namespaces

**Comment by lrusak** [comment]
any objections?

File: xbmc/cores/omxplayer/OMXImage.cpp [hunk 8]
```diff
@@ -11,7 +11,7 @@
 #include "ServiceBroker.h"
 #include "URL.h"
 #include "utils/log.h"
-#include "platform/posix/XMemUtils.h"
+#include "utils/MemUtils.h"
```
Comment (notspiff) [inline xbmc/cores/omxplayer/OMXImage.cpp]: if this is the only change required, this header is likely unused?

File: xbmc/guilib/Texture.h [hunk 9]
```diff
@@ -10,9 +10,7 @@
 
 #include "XBTF.h"
 #include "guilib/imagefactory.h"
-#ifdef TARGET_POSIX
-#include "platform/posix/XMemUtils.h"
-#endif
+#include "utils/MemUtils.h"
```
Comment (notspiff) [inline xbmc/guilib/Texture.h]: ditto.

File: xbmc/platform/linux/OMXCore.cpp [hunk 10]
```diff
@@ -11,14 +11,11 @@
 
 #include "OMXCore.h"
 #include "utils/log.h"
+#include "utils/MemUtils.h"
```
Comment (Rechi) [inline xbmc/platform/linux/OMXCore.cpp]: same here

**Comment by lrusak** [comment]
updated

**Review by Rechi — APPROVED** [review]

**Pull request merged**